### PR TITLE
PP-4096 update service name

### PR DESCRIPTION
**Pull request by georgeracu**
* Updated `ServiceEntity#addServiceName` to act as an add or update if
    exists. This is necessary when applying the patch operations and in the end
    the entity is persisted to the db.
    * Updated `ServiceNameEntity` to be created from `ServiceUpdateRequest`,
    such that when a PATCH operation is received for service name, then it
    can be created successfully
    * Added `ServiceUpdatedRequest#getUpdateRequests` to return a list of
    updated requests from the payload, so the PATCH payload can be deserialised
    into individual `ServiceUpdateRequests`
    * Updated `ServiceRequestValidator` to validate for the service_name object as well
    * Renamed `ServiceRequestValidator#FIELD_NAME` not to clash with the
    actual field `FIELD_SERVICE_NAME`
    * Added new validation methods in RequestValidations that will return
    an empty collection if no errors, instead of wrapping the collection into
    an `Optional` and then testing if the optional is present. Unnecessary
    wrapping into `Optional`
    * Added unit tests for the new resource and the new validator
    * Added JSON files where to keep the payload for testing the request or
    the validator. It feels easier to see the JSON object in a separate file
    than to create it by hand using lists, maps etc. and to hope that you got
    it right first when you try to serialize it into JSON. This is a pattern
    we might adopt in other places
    * Updated `ServiceUpdateRequest` to handle both arrays and objects for the
    PATCH request
    * Removed the `name` field from the `equals` and `hash` methods from a
    `ServiceNameEntity` such that when adding to the set of service names from
    `ServiceEntity` then the `equals` method will reflect that a service can
    have only one service name for a given language. This is also useful in
    `addOrUpdate` when if the service name for the given language already
    exists then it will be updated

### CODE REVIEW

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+        Response response = resources.target(format(API_PATH, externalId))
+                .request()
+                .method("PATCH", Entity.json(jsonPayload));
+
+        assertThat(response.getStatus(), is(404));
+    }
+
+    @Test
+    public void shouldError400_ifFieldValueMissing() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: This is equivalent to `ServiceResourceUpdateTest.shouldError400_ifMandatoryFieldMissing()`

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        assertThat(response.getStatus(), is(200));
+
+        String body = response.readEntity(String.class);
+        JsonPath json = JsonPath.from(body);
+
+        assertThat(json.getList("gateway_account_ids"), contains("1014748185"));
+    }
+
+    @Test
+    public void shouldSuccess_whenAddGatewayAccountIds_whereThereIsGatewayAccountIds() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: This is the equivalent of `ServiceResourceUpdateTest.shouldSuccess_whenAddGatewayAccountIds()`

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        assertThat(response.getStatus(), is(200));
+
+        String body = response.readEntity(String.class);
+        JsonPath json = JsonPath.from(body);
+
+        assertThat(json.getList("gateway_account_ids"), containsInAnyOrder("1014748185", gatewayAccountId));
+    }
+
+    @Test
+    public void shouldReturn409_whenAddGatewayAccountIds_andGatewayAccountId_isUsedByAnotherService() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: This is the equivalent of `ServiceUpdateTest.shouldError409_whenAddGatewayAccountIds_ifGatewayAccountIdAlreadyUsedByAnotherService()`

File: src/main/java/uk/gov/pay/adminusers/persistence/entity/service/ServiceNameEntity.java [hunk 4]
```diff
@@ -89,13 +105,12 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) return false;
         ServiceNameEntity that = (ServiceNameEntity) o;
         return Objects.equals(service, that.service) &&
-                Objects.equals(language, that.language) &&
-                Objects.equals(name, that.name);
+                Objects.equals(language, that.language);
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/persistence/entity/service/ServiceNameEntity.java]: 👍

File: src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java [hunk 5]
```diff
@@ -48,41 +57,59 @@ public ServiceRequestValidator(RequestValidations requestValidations) {
...
-        if (errors.isPresent()) {
-            return Optional.of(Errors.from(errors.get()));
+        List<String> finalErrors = new ArrayList<>();
+        ArrayNode operations = new ObjectMapper().createArrayNode();
+        try {
+            operations = (ArrayNode) new ObjectMapper().readTree(payload.toString());
+        } catch (IOException e) {
+            LOGGER.info("There was an exception processing update request [{}]", e.getMessage());
+            finalErrors.add("There was an error processing update");
+        } catch (ClassCastException e) {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java]: 👍

File: src/main/java/uk/gov/pay/adminusers/validations/RequestValidations.java [hunk 6]
```diff
@@ -29,10 +29,18 @@
         return applyCheck(payload, notExistOrEmpty(), fieldNames, "Field [%s] is required");
     }
 
+    public List<String> checkIfExistsOrEmptyV2(JsonNode payload, String... fieldNames) {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/validations/RequestValidations.java]: I will reverto to the old way, not enough time to change everything 😞

File: src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java [hunk 7]
```diff
@@ -122,4 +149,11 @@ private void validateMerchantEmail(JsonNode payload) throws ValidationException
         return Optional.empty();
     }
 
+    private List<String> checkIfValidJson(JsonNode payload) {
+        if (payload == null || !payload.isObject()) {
+            return Collections.singletonList(format("Value for path [%s] must be a JSON", FIELD_CUSTOM_BRANDING));
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java]: Refactored to get a field name as well

File: src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java [hunk 8]
```diff
@@ -48,41 +57,59 @@ public ServiceRequestValidator(RequestValidations requestValidations) {
...
+            }
 
-    private Optional<List<String>> checkIfNotEmptyAndJson(JsonNode payload) {
-        if (payload == null || !payload.isObject()) {
-            return Optional.of(Collections.singletonList(format("Value for path [%s] must be a JSON", FIELD_CUSTOM_BRANDING)));
+            String op = item.get("op").asText();
+            if (!VALID_ATTRIBUTE_UPDATE_OPERATIONS.get(path).contains(op)) {
+                finalErrors.add(format("Operation [%s] is invalid for path [%s]", op, path));
+            }
+        });
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java]: split it into smaller methods

File: src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java [hunk 9]
```diff
@@ -48,41 +57,59 @@ public ServiceRequestValidator(RequestValidations requestValidations) {
...
+                finalErrors.addAll(requestValidations.checkIfExistsOrEmptyV2(item, FIELD_VALUE));
+                if (finalErrors.isEmpty()) {
+                    finalErrors.addAll(requestValidations.checkMaxLengthV2(item, SERVICE_NAME_MAX_LENGTH, FIELD_VALUE));
+                }
+            }
+        });
+        if (!finalErrors.isEmpty()) {
+            return Optional.of(Errors.from(finalErrors));
         }
+        operations.forEach(item -> {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/main/java/uk/gov/pay/adminusers/resources/ServiceRequestValidator.java]: I kept the old pattern where we apply a check at a time and fail early if errors.

File: src/test/java/uk/gov/pay/adminusers/persistence/entity/service/ServiceNameEntityTest.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        payload.put("op", "replace");
+        payload.put("path", "name");
+        payload.put("value", ImmutableMap.of("en", "new-en-name"));
+
+        final ObjectMapper mapper = new ObjectMapper();
+        final String stringPayload = mapper.writeValueAsString(payload);
+        ServiceUpdateRequest serviceUpdateRequest = ServiceUpdateRequest.from(mapper.readTree(stringPayload));
+
+        ServiceNameEntity serviceNameEntity = ServiceNameEntity.from(serviceUpdateRequest);
+        assertThat(serviceNameEntity, is(notNullValue()));
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/persistence/entity/service/ServiceNameEntityTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/resources/ServiceRequestValidatorTest.java [hunk 11]
```diff
@@ -25,21 +28,34 @@
     private ServiceRequestValidator serviceRequestValidator = new ServiceRequestValidator(new RequestValidations());
 
     @Test
-    public void shouldSuccess_whenUpdate_whenAllFieldPresentAndValid() {
+    public void shouldSuccess_whenUpdate_whenAllFieldPresentAndValid_andJsonNotArray() {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/resources/ServiceRequestValidatorTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/resources/ServiceRequestValidatorTest.java [hunk 12]
```diff
@@ -25,21 +28,34 @@
...
         ImmutableMap<String, String> payload = ImmutableMap.of("path", "name", "op", "replace", "value", "example-name");
 
         Optional<Errors> errors = serviceRequestValidator.validateUpdateAttributeRequest(mapper.valueToTree(payload));
         assertFalse(errors.isPresent());
     }
 
+    @Test
+    public void shouldSuccess_whenUpdateArray_withAllFieldsPresentAndValid() throws IOException {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/resources/ServiceRequestValidatorTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 13]
```diff
@@ -189,4 +191,24 @@ public void shouldError_IfAGatewayAccountAlreadyAssignedToAService() throws Exce
...
+        ServiceEntity serviceEntity = mock(ServiceEntity.class);
+
+        when(serviceDao.findByExternalId(serviceId)).thenReturn(Optional.of(serviceEntity));
+        when(serviceEntity.toService()).thenReturn(Service.from());
+
+        Optional<Service> maybeService = updater.doUpdate(serviceId, request);
+
+        assertThat(maybeService.isPresent(), is(true));
+        ServiceNameEntity serviceNameEntity = ServiceNameEntity.from(SupportedLanguage.WELSH, nameToUpdate);
+        verify(serviceEntity, times(1)).addOrUpdateServiceName(serviceNameEntity);
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: Now `name` is added to the `equals` and `hash` of `ServiceNameEntity`

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 14]
```diff
@@ -189,4 +191,24 @@ public void shouldError_IfAGatewayAccountAlreadyAssignedToAService() throws Exce
...
+        verify(serviceDao, times(1)).merge(serviceEntity);
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: Well spotted, added `inOrder` for the methods on the dao

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+                    mockedServicesFactory))
+            .build();
+
+    @Before
+    public void setUp() {
+        when(mockedServicesFactory.serviceUpdater()).thenReturn(serviceUpdater);
+    }
+
+    @Test
+    public void shouldSuccess_whenUpdateOnlyServiceName() {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: Renamed to the old name

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+        Response response = resources.target(format(API_PATH, thisServiceEntity.getExternalId()))
+                .request()
+                .method("PATCH", Entity.json(jsonPayload));
+
+        assertThat(response.getStatus(), is(200));
+
+        String body = response.readEntity(String.class);
+        JsonPath json = JsonPath.from(body);
+
+        assertThat(json.get("name"), is(thisServiceEntity.getName()));
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: Updated assertion

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+
+        String body = response.readEntity(String.class);
+        JsonPath json = JsonPath.from(body);
+
+        assertThat(json.get("name"), is("System Generated"));
+        assertCyServiceNameJson("new-cy-name", json);
+    }
+
+    @Test
+    public void shouldError404_ifServiceExternalIdDoesNotExist() {
```
[the remark this answers is not in the document]
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: Pointed fixture to `update-name-only.json` which has the same payload as the previous test

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java (hunk 1 above)
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: I renamed the tests that are asserting for missing fields to be explicit that is mandatory

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java (hunk 2 above)
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java (hunk 3 above)
Comment (georgeracu) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: 👍

File: src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexbishop1) [reply src/test/java/uk/gov/pay/adminusers/unit/service/ServiceResourceUpdateTest.java]: Sorry, with the comments saying, “This is the equivalent of the old…”, I wasn’t asking you to rename them, just keeping notes (for myself really) of what had been migrated!

File: src/main/java/uk/gov/pay/adminusers/resources/ServiceResource.java [hunk 18]
```diff
@@ -152,10 +152,18 @@ public Response updateServiceAttribute(@PathParam("serviceExternalId") String se
...
+                .orElseGet(() -> processPayload(serviceExternalId, payload));
+    }
 
+    private Response processPayload(String serviceExternalId, JsonNode payload) {
+        try {
+            final List<ServiceUpdateRequest> requests = ServiceUpdateRequest.getUpdateRequests(payload);
+            return serviceServicesFactory.serviceUpdater().doUpdate(serviceExternalId, requests)
+                    .map(service -> Response.status(OK).entity(service).build())
+                    .orElseGet(() -> Response.status(NOT_FOUND).build());
+        } catch (ValidationException e) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/adminusers/resources/ServiceResource.java]: It seems the convention for other endpoints in this resource is to just let the exception propagate upwards

File: src/test/java/uk/gov/pay/adminusers/model/ServiceUpdateRequestTest.java [hunk 19]
```diff
@@ -32,4 +36,46 @@ public void shouldTransformToObjectCorrectly() throws Exception {
...
+                "{\n" +
+                "  \"op\": \"replace\",\n" +
+                "  \"path\": \"name\",\n" +
+                "  \"value\": \"new-en-name\"\n" +
+                "}\n";
+
+        final List<ServiceUpdateRequest> requests = ServiceUpdateRequest.getUpdateRequests(new ObjectMapper().readTree(jsonPayload));
+
+        assertThat(requests.size(), is(1));
+        assertThat(requests.get(0).getPath(), is("name"));
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/model/ServiceUpdateRequestTest.java]: Could check the `"value"` here as well, I suppose

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 20]
```diff
@@ -189,4 +193,27 @@ public void shouldError_IfAGatewayAccountAlreadyAssignedToAService() throws Exce
...
+        when(serviceDao.findByExternalId(serviceId)).thenReturn(Optional.of(serviceEntity));
+        when(serviceEntity.toService()).thenReturn(Service.from());
+
+        InOrder inOrder = Mockito.inOrder(serviceDao);
+
+        Optional<Service> maybeService = updater.doUpdate(serviceId, request);
+
+        assertThat(maybeService.isPresent(), is(true));
+        ServiceNameEntity serviceNameEntity = ServiceNameEntity.from(SupportedLanguage.WELSH, nameToUpdate);
+        inOrder.verify(serviceDao, times(1)).findByExternalId(serviceId);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: I don’t think we really need to verify this; it’s covered by the stubbing earlier

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 21]
```diff
@@ -189,4 +193,27 @@ public void shouldError_IfAGatewayAccountAlreadyAssignedToAService() throws Exce
...
+        inOrder.verify(serviceDao, times(1)).merge(serviceEntity);
+        verify(serviceEntity, times(1)).addOrUpdateServiceName(serviceNameEntity);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: Ah, I think I wasn’t clear. I meant to do something like this:

```java
InOrder inOrder = Mockito.inOrder(serviceEntity, serviceDao);
// …
inOrder.verify(serviceEntity).addOrUpdateServiceName(serviceNameEntity);
inOrder.verify(serviceDao).merge(serviceEntity);
```

We want to ensure the `serviceEntity` is updated before it’s merged via the `serviceDao`.

Also, `times(1)` isn’t needed: it’s the default.

**Comment by alexbishop1** [comment]
The answer to what happens if we get an English service name update where both "path": "name" and "path": "service_name", "value": {"en": …} operations are specified but the names conflict is that one will override the other. This should never happen in practice though (only selfservice will call this endpoint), so we don’t really need to worry about it.

**Pull request merged**